mathics/algorithm/clusters.py

```diff
@@ -0,0 +1,714 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+from __future__ import division
+
+import random
+from heapq import nsmallest
+from itertools import chain
+import bisect
+import math
+
+# publications used for this file:
+
+# [Kurita1991] Takito Kurita: "An Efficient Agglomerative Clustering Algorithm Using A Heap", Pattern Recognition,
+# Volume 24 Issue 3, 1991.
+# [Ng1994] Raymond T. Ng, Jiawei Han: Efficient and Effective Clustering Methods for Spatial Data Mining.
+# VLDB 1994: 144-155
+# [Rangel2016] E. Rangel, W. Hendrix, A. Agrawal, W.-keng Liao, and A. Choudhary, “AGORAS: A Fast Algorithm for
+# Estimating Medoids in Large Datasets,” in Proceedings of International Conference on Computational Science (ICCS),
+# 2016.
+# [Hamerly2003] Greg Hamerly, Charles Elkan, Learning the k in k-means. In proceedings of the seventeenth
+# annual conference on neural information processing systems (NIPS), pages 281-288, December 2003.
+# [Desgraupes2013] Bernard Desgraupes, Package clusterCrit for R: Clustering Indices, available online at
+# https://cran.r-project.org/web/packages/clusterCrit/vignettes/clusterCrit.pdf
+# [Greutert2003] Andreas Greutert. Methoden zur Schätzung der Clusteranzahl. Ausgabe: 29. Oktober 2003.
+# Diplomarbeit. ETH Zürich.
+
+# for agglomerative clustering, we use [Kurita1991].
+
+# for hierarchical clustering, we use ClaraNS (see [Ng1994]), but we start local searches not with a random sample,
+# but with a seed computed via AGORAS (see [Rangel2016]). for clustering without knowing the number of clusters k,
+# we use the approach described in [Hamerly2003]. clustering quality is measures using McClain-Rao's measure (see
+# [Desgraupes2013]).
+
+# we modify ClaraNS by keeping track of the randomly picked swaps and not trying the same swap twice if no
+# improvement was made in the meantime.
+
+
+# the following value defines by how much the McClain-Rao limit (for considering a clustering finished) changes
+# with each binary split of clustering. this is an experimental value.
+_mcclain_rao_limit_factor = 10.
```

Comment:
The assumption of the current algorithm on cluster count detection is this: for each possible clustering, we compute the ratio of the average in-cluster distance (i.e. the distances from all points of a cluster to the cluster's center) to the average between-cluster distance (the distance between cluster centers). With each split, that doubles the count of clusters, this ratio has to lower by a constant factor, or we won't do the split. For example, if we go from 2 to 4 clusters, then the ratio of each of the 4 clusters has to be smaller than the 2-cluster-ratio divided by some `x`. This `x` is `_mcclain_rao_limit_factor`. It basically says: if we split into more clusters, and the configuration isn't more compact (i.e. centered around cluster points) by this and that degree, we don't split, since then these new clusters are not delineated properly. So, to summarize, a `_mcclain_rao_limit_factor` of 10 states that the distance ratio needs to diminish by factor of 10 with each split, but we do not care what the absolute value of the distance ratio is.

Even though `_mcclain_rao_limit_factor` does not depend on the input data, it does obviously influence the granularity of the result. Maybe an option `Granularity` would be nice?